# 项目：整理Netflix电影演员评分数据

## 分析目标

此数据分析的目的是，整理不同流派影视作品，比如喜剧片、动作片、科幻片中，各演员出演作品的平均IMDB评分，从而挖掘出各个流派中的高评分作品演员。

本实战项目的目的在于练习整理数据，从而得到可供下一步分析的数据。

## 简介

原始数据集记录了截止至2022年7月美国地区可观看的所有Netflix电视剧及电影数据。数据集包含两个数据表：`titles.csv`和`credits.csv`。

`titles.csv`包含电影及电视剧相关信息，包括影视作品ID、标题、类型、描述、流派、IMDB（一个国外的在线评分网站）评分，等等。`credits.csv`包含超过7万名出现在Netflix影视作品的导演及演员信息，包括名字、影视作品ID、人物名、演职员类型（导演/演员）等。

`titles.csv`每列的含义如下：
- id：影视作品ID。
- title：影视作品标题。
- show_type：作品类型，电视节目或电影。
- description：简短描述。
- release_year：发布年份。
- age_certification：适龄认证。
- runtime：每集电视剧或电影的长度。
- genres：流派类型列表。
- production_countries：出品国家列表。
- seasons：如果是电视剧，则是季数。
- imdb_id：IMDB的ID。
- imdb_score：IMDB的评分。
- imdb_votes：IMDB的投票数。
- tmdb_popularity：TMDB的流行度。
- tmdb_score：TMDB的评分。

`credits.csv`每列的含义如下：
- person_ID：演职员ID。
- id：参与的影视作品ID。
- name：姓名。
- character_name：角色姓名。
- role：演职员类型，演员或导演。

In [1]:
import pandas as pd
import numpy as np
pd.set_option("display.max_columns",50)
pd.set_option("display.max_colwidth",20)

In [2]:
df1 = pd.read_csv("titles.csv")
df1.head(20)

id                title   type          description  release_year  \
0   ts300399  Five Came Back: ...   SHOW  This collection ...          1945   
1    tm84618          Taxi Driver  MOVIE  A mentally unsta...          1976   
2   tm154986          Deliverance  MOVIE  Intent on seeing...          1972   
3   tm127384  Monty Python and...  MOVIE  King Arthur, acc...          1975   
4   tm120801      The Dirty Dozen  MOVIE  12 American mili...          1967   
5    ts22164  Monty Python's F...   SHOW  A British sketch...          1969   
6    tm70993        Life of Brian  MOVIE  Brian Cohen is a...          1979   
7    tm14873          Dirty Harry  MOVIE  When a madman du...          1971   
8   tm119281     Bonnie and Clyde  MOVIE  In the 1930s, bo...          1967   
9    tm98978      The Blue Lagoon  MOVIE  Two small childr...          1980   
10   tm44204  The Guns of Nava...  MOVIE  A team of allied...          1961   
11   tm67378    The Professionals  MOVIE  An arrogant Texa...          1966   
12   tm69997  Richard Pryor: L...  MOVIE  Richard Pryor de...          1979   
13   tm16479      White Christmas  MOVIE  Two talented son...          1954   
14  tm135083        Cairo Station  MOVIE  Qinawi, a physic...          1958   
15   tm89386     Hitler: A Career  MOVIE  A keen chronicle...          1977   
16  tm156453                  FTA  MOVIE  A documentary ab...          1972   
17   tm14350     Alexandria… Why?  MOVIE  Set against the ...          1979   
18   tm81728             The Land  MOVIE  Set in 1933, the...          1969   
19   tm94651              Dostana  MOVIE  Two close friend...          1980   

   age_certification  runtime               genres production_countries  \
0              TV-MA       51    ['documentation']               ['US']   
1                  R      114   ['drama', 'crime']               ['US']   
2                  R      109  ['drama', 'actio...               ['US']   
3                 PG       91  ['fantasy', 'act...               ['GB']   
4                NaN      150    ['war', 'action']         ['GB', 'US']   
5              TV-14       30  ['comedy', 'euro...               ['GB']   
6                  R       94           ['comedy']               ['GB']   
7                  R      102  ['thriller', 'ac...               ['US']   
8                  R      110  ['crime', 'drama...               ['US']   
9                  R      104  ['romance', 'act...               ['US']   
10               NaN      158  ['action', 'dram...         ['GB', 'US']   
11             PG-13      117  ['western', 'act...               ['US']   
12                 R       78  ['comedy', 'docu...               ['US']   
13               NaN      115  ['romance', 'com...               ['US']   
14               NaN       77  ['drama', 'crime...               ['EG']   
15                PG      150  ['history', 'doc...               ['DE']   
16                 R       97  ['war', 'documen...               ['US']   
17               NaN      133            ['drama']               ['EG']   
18               NaN      129            ['drama']               ['EG']   
19               NaN      161  ['drama', 'comed...               ['IN']   

    seasons    imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
0       1.0        NaN         NaN         NaN            0.600         NaN  
1       NaN  tt0075314         8.2    808582.0           40.965       8.179  
2       NaN  tt0068473         7.7    107673.0           10.010       7.300  
3       NaN  tt0071853         8.2    534486.0           15.461       7.811  
4       NaN  tt0061578         7.7     72662.0           20.398       7.600  
5       4.0  tt0063929         8.8     73424.0           17.617       8.306  
6       NaN  tt0079470         8.0    395024.0           17.770       7.800  
7       NaN  tt0066999         7.7    155051.0           12.817       7.500  
8       NaN  tt0061418         7.7    112048.0           15.687       7.

In [3]:
df2 = pd.read_csv("credits.csv")
df2.sample(20)

person_id         id                 name            character  \
10081     296144    tm44378         Sergio Monge               Tony 1   
77210    2344967   tm863829       Amalia Suryani  Indonesian Friend 2   
45135     654771   tm817504  Earnestine Phillips               Esther   
38176     279919   tm358666        Idit Teperson  Malka, Rabbi's wife   
68105     852365   tm413169        Will Reichelt  JW Rooster III (...   
59504    1289880   tm873159           Ju Xiaowen        The Eye Demon   
27841      60457   tm232797        Lee Armstrong            Grenadier   
62125    1154847   tm810357        Harold Harris                  NaN   
63405       5271   tm942492         Vincent Tong  Drac Shadows & N...   
34103     374905   tm420334       Amber Hodgkiss                Ginny   
45055     941718    ts82913        Ethan Herisse   Young Yusef Salaam   
36594     656827   tm311789    David Henry Begin      Young Jon Garth   
184        11489    tm70993        Chris Langham  Alfonso / Giggli...   
41919     590079   tm432313    Chandini Chowdary                Neela   
74727      62248   tm960430         Ko Shibasaki  Yayoi Hayama (vo...   
59417      31702    ts89330          Ayumu Saitô  Mamoru Jōgasaki（...   
73342     286039  tm1132325       George Mariyan                  NaN   
51016    1080424   tm936396   Patrick Lamont Jr.                   DJ   
2772       33667   tm191989         Antone Pagán  Prison Guard for...   
48881       3777   tm444247   Chloë Grace Moretz              Herself   

           role  
10081     ACTOR  
77210     ACTOR  
45135     ACTOR  
38176     ACTOR  
68105     ACTOR  
59504     ACTOR  
27841     ACTOR  
62125  DIRECTOR  
63405     ACTOR  
34103     ACTOR  
45055     ACTOR  
36594     ACTOR  
184       ACTOR  
41919     ACTOR  
74727     ACTOR  
59417     ACTOR  
73342     ACTOR  
51016     ACTOR  
2772      ACTOR  
48881     ACTOR

**评估数据**

#### 结构性问题

#### 某些列如作品的类别是一个列表，存在多值，应该是要分行的在title里

In [4]:
df1["genres"][1]  #可以见到表面是一个列表，实际是一个字符串，这里会用到eval函数，传入字符串会执行字符串内的内容

"['drama', 'crime']"

In [5]:
df1["genres"] = df1["genres"].apply(lambda s : eval(s))  #相当于去掉外面那层引号

In [6]:
df1

id                title   type          description  \
0      ts300399  Five Came Back: ...   SHOW  This collection ...   
1       tm84618          Taxi Driver  MOVIE  A mentally unsta...   
2      tm154986          Deliverance  MOVIE  Intent on seeing...   
3      tm127384  Monty Python and...  MOVIE  King Arthur, acc...   
4      tm120801      The Dirty Dozen  MOVIE  12 American mili...   
...         ...                  ...    ...                  ...   
5845  tm1014599            Fine Wine  MOVIE  A beautiful love...   
5846   tm898842          C/O Kaadhal  MOVIE  A heart warming ...   
5847  tm1059008              Lokillo  MOVIE  A controversial ...   
5848  tm1035612  Dad Stop Embarra...  MOVIE  Jamie Foxx, Davi...   
5849   ts271048  Mighty Little Bh...   SHOW  With winter behi...   

      release_year age_certification  runtime               genres  \
0             1945             TV-MA       51      [documentation]   
1             1976                 R      114       [drama, crime]   
2             1972                 R      109  [drama, action, ...   
3             1975                PG       91  [fantasy, action...   
4             1967               NaN      150        [war, action]   
...            ...               ...      ...                  ...   
5845          2021               NaN      100     [romance, drama]   
5846          2021               NaN      134              [drama]   
5847          2021               NaN       90             [comedy]   
5848          2021             PG-13       37                   []   
5849          2021               NaN        7  [family, animati...   

     production_countries  seasons     imdb_id  imdb_score  imdb_votes  \
0                  ['US']      1.0         NaN         NaN         NaN   
1                  ['US']      NaN   tt0075314         8.2    808582.0   
2                  ['US']      NaN   tt0068473         7.7    107673.0   
3                  ['GB']      NaN   tt0071853         8.2    534486.0   
4            ['GB', 'US']      NaN   tt0061578         7.7     72662.0   
...                   ...      ...         ...         ...         ...   
5845               ['NG']      NaN  tt13857480         6.8        45.0   
5846                   []      NaN  tt11803618         7.7       348.0   
5847               ['CO']      NaN  tt14585902         3.8        68.0   
5848               ['US']      NaN         NaN         NaN         NaN   
5849                   []      1.0  tt13711094         7.8        18.0   

      tmdb_popularity  tmdb_score  
0               0.600         NaN  
1              40.965       8.179  
2              10.010       7.300  
3              15.461       7.811  
4              20.398       7.600  
...               ...         ...  
5845            1.466         NaN  
5846              NaN         NaN  
5847           26.005       6.300  
5848            1.296      10.000  
5849            2.289      10.000  

[5850 rows x 15 columns]

In [7]:
#### 拆行用explode方法

In [8]:
df1 = df1.explode("genres")
df1.head(10)

id                title   type          description  release_year  \
0  ts300399  Five Came Back: ...   SHOW  This collection ...          1945   
1   tm84618          Taxi Driver  MOVIE  A mentally unsta...          1976   
1   tm84618          Taxi Driver  MOVIE  A mentally unsta...          1976   
2  tm154986          Deliverance  MOVIE  Intent on seeing...          1972   
2  tm154986          Deliverance  MOVIE  Intent on seeing...          1972   
2  tm154986          Deliverance  MOVIE  Intent on seeing...          1972   
2  tm154986          Deliverance  MOVIE  Intent on seeing...          1972   
3  tm127384  Monty Python and...  MOVIE  King Arthur, acc...          1975   
3  tm127384  Monty Python and...  MOVIE  King Arthur, acc...          1975   
3  tm127384  Monty Python and...  MOVIE  King Arthur, acc...          1975   

  age_certification  runtime         genres production_countries  seasons  \
0             TV-MA       51  documentation               ['US']      1.0   
1                 R      114          drama               ['US']      NaN   
1                 R      114          crime               ['US']      NaN   
2                 R      109          drama               ['US']      NaN   
2                 R      109         action               ['US']      NaN   
2                 R      109       thriller               ['US']      NaN   
2                 R      109       european               ['US']      NaN   
3                PG       91        fantasy               ['GB']      NaN   
3                PG       91         action               ['GB']      NaN   
3                PG       91         comedy               ['GB']      NaN   

     imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
0        NaN         NaN         NaN            0.600         NaN  
1  tt0075314         8.2    808582.0           40.965       8.179  
1  tt0075314         8.2    808582.0           40.965       8.179  
2  tt0068473         7.7    107673.0           10.010       7.300  
2  tt0068473         7.7    107673.0           10.010       7.300  
2  tt0068473         7.7    107673.0           10.010       7.300  
2  tt0068473         7.7    107673.0           10.010       7.300  
3  tt0071853         8.2    534486.0           15.461       7.811  
3  tt0071853         8.2    534486.0           15.461       7.811  
3  tt0071853         8.2    534486.0           15.461       7.811

In [9]:
#production_countries 也是需要拆的
df1["production_countries"][0]

"['US']"

In [10]:
df1["production_countries"] = df1["production_countries"].apply(lambda s : eval(s))

In [11]:
df1["production_countries"][0]

['US']

In [12]:
df1 = df1.explode("production_countries")
df1

id                title   type          description  \
0      ts300399  Five Came Back: ...   SHOW  This collection ...   
1       tm84618          Taxi Driver  MOVIE  A mentally unsta...   
1       tm84618          Taxi Driver  MOVIE  A mentally unsta...   
2      tm154986          Deliverance  MOVIE  Intent on seeing...   
2      tm154986          Deliverance  MOVIE  Intent on seeing...   
...         ...                  ...    ...                  ...   
5847  tm1059008              Lokillo  MOVIE  A controversial ...   
5848  tm1035612  Dad Stop Embarra...  MOVIE  Jamie Foxx, Davi...   
5849   ts271048  Mighty Little Bh...   SHOW  With winter behi...   
5849   ts271048  Mighty Little Bh...   SHOW  With winter behi...   
5849   ts271048  Mighty Little Bh...   SHOW  With winter behi...   

      release_year age_certification  runtime         genres  \
0             1945             TV-MA       51  documentation   
1             1976                 R      114          drama   
1             1976                 R      114          crime   
2             1972                 R      109          drama   
2             1972                 R      109         action   
...            ...               ...      ...            ...   
5847          2021               NaN       90         comedy   
5848          2021             PG-13       37            NaN   
5849          2021               NaN        7         family   
5849          2021               NaN        7      animation   
5849          2021               NaN        7         comedy   

     production_countries  seasons     imdb_id  imdb_score  imdb_votes  \
0                      US      1.0         NaN         NaN         NaN   
1                      US      NaN   tt0075314         8.2    808582.0   
1                      US      NaN   tt0075314         8.2    808582.0   
2                      US      NaN   tt0068473         7.7    107673.0   
2                      US      NaN   tt0068473         7.7    107673.0   
...                   ...      ...         ...         ...         ...   
5847                   CO      NaN  tt14585902         3.8        68.0   
5848                   US      NaN         NaN         NaN         NaN   
5849                  NaN      1.0  tt13711094         7.8        18.0   
5849                  NaN      1.0  tt13711094         7.8        18.0   
5849                  NaN      1.0  tt13711094         7.8        18.0   

      tmdb_popularity  tmdb_score  
0               0.600         NaN  
1              40.965       8.179  
1              40.965       8.179  
2              10.010       7.300  
2              10.010       7.300  
...               ...         ...  
5847           26.005       6.300  
5848            1.296      10.000  
5849            2.289      10.000  
5849            2.289      10.000  
5849            2.289      10.000  

[17818 rows x 15 columns]

In [13]:
#  title文件处理完了结构性问题，我们来看credit的  credit符合每行一个观察值，每列一个变量，并且单元格只有一个变量的要求

In [14]:
df2

person_id         id                name            character      role
0           3748    tm84618      Robert De Niro        Travis Bickle     ACTOR
1          14658    tm84618        Jodie Foster        Iris Steensma     ACTOR
2           7064    tm84618       Albert Brooks                  Tom     ACTOR
3           3739    tm84618       Harvey Keitel  Matthew 'Sport' ...     ACTOR
4          48933    tm84618     Cybill Shepherd                Betsy     ACTOR
...          ...        ...                 ...                  ...       ...
77796     736339  tm1059008    Adelaida Buscato            María Paz     ACTOR
77797     399499  tm1059008  Luz Stella Luengas         Karen Bayona     ACTOR
77798     373198  tm1059008         Inés Prieto                Fanny     ACTOR
77799     378132  tm1059008        Isabel Gaona               Cacica     ACTOR
77800    1950416  tm1059008      Julian Gaviria                  NaN  DIRECTOR

[77801 rows x 5 columns]

### 数据干净度

In [15]:
#titile文件的数据干净度我们来查看一下

In [16]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
Index: 17818 entries, 0 to 5849
Data columns (total 15 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   id                    17818 non-null  object 
 1   title                 17817 non-null  object 
 2   type                  17818 non-null  object 
 3   description           17790 non-null  object 
 4   release_year          17818 non-null  int64  
 5   age_certification     10889 non-null  object 
 6   runtime               17818 non-null  int64  
 7   genres                17755 non-null  object 
 8   production_countries  17439 non-null  object 
 9   seasons               6224 non-null   float64
 10  imdb_id               17116 non-null  object 
 11  imdb_score            16976 non-null  float64
 12  imdb_votes            16945 non-null  float64
 13  tmdb_popularity       17663 non-null  float64
 14  tmdb_score            17241 non-null  float64
dtypes: float64(5), int64(2), 

从输出结果来看，`cleaned_titles`数据共有17818条观察值，`title`、`description`、`age_certification`、`genres`、`production_countries`、`seasons`、`imdb_id`、`imdb_score`、`tmdb_popularity`、`tmdb_score`、`imdb_votes`、`tmdb_popularity`、`tmdb_score`变量均存在缺失值，将在后续进行评估和清理。

此外，`release_year`表示年份，数据类型不应为数字，应为日期，所以需要进行数据格式转换。

In [17]:
df1["release_year"] = pd.to_datetime(df1["release_year"],format = '%Y')  
#代码中的 format='%Y' 表示输入的 release_year 列中的数据仅包含四位数的年份
df1["release_year"]

0      1945-01-01
1      1976-01-01
1      1976-01-01
2      1972-01-01
2      1972-01-01
          ...    
5847   2021-01-01
5848   2021-01-01
5849   2021-01-01
5849   2021-01-01
5849   2021-01-01
Name: release_year, Length: 17818, dtype: datetime64[ns]

In [18]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 77801 entries, 0 to 77800
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   person_id  77801 non-null  int64 
 1   id         77801 non-null  object
 2   name       77801 non-null  object
 3   character  68029 non-null  object
 4   role       77801 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.0+ MB


从输出结果来看，`cleaned_credits`数据共有77801条观察值，其中`character`变量存在缺失值，将在后续进行评估和清理。

此外，`person_id`表示演职员ID，数据类型不应为数字，应为字符串，所以需要进行数据格式转换。

In [19]:
df2["person_id"]  = df2["person_id"].astype(str)
df2["person_id"]

0           3748
1          14658
2           7064
3           3739
4          48933
          ...   
77796     736339
77797     399499
77798     373198
77799     378132
77800    1950416
Name: person_id, Length: 77801, dtype: object

**接下来处理空缺值**

在`cleaned_titles`中，`title`、`description`、`age_certification`、`genres`、`production_countries`、`seasons`、`imdb_id`、`imdb_score`、`tmdb_popularity`、`tmdb_score`、`imdb_votes`、`tmdb_popularity`、`tmdb_score`变量存在缺失值。

由于影视作品的标题、描述、适龄认证、发行国家、电视剧季数、IMDB的ID、TMDB的流行度、TMDB的评分，并不影响我们挖掘各个流派中的高IMDB评分作品演员，所以可以保留`title`、`description`、`age_certification`、`production_countries`、`seasons`、`imdb_id`、`tmdb_popularity`、`tmdb_score`、`imdb_votes`、`tmdb_popularity`、`tmdb_score`变量值存在空缺的观察值。

但`imdb_score`和`genres`，即IMDB评分和流派，和我们后续要做的分析息息相关。

先提取出`imdb_score`缺失观察值进行查看。

In [20]:
## 因为我们整个是要获取  所有流派中评分最高的角色，因此，角色、评分、流派才是最重要的

In [21]:
df1.query("imdb_score.isnull()")  #记住了query也是筛选行，但是里面是整个字符串，列名不需要再给影号了

id                title   type          description release_year  \
0      ts300399  Five Came Back: ...   SHOW  This collection ...   1945-01-01   
75     tm132164  Bill Hicks: Sane...  MOVIE  Sane Man was fil...   1989-01-01   
145    ts251477      My First Errand   SHOW  “Hajimete no Ots...   1991-01-01   
145    ts251477      My First Errand   SHOW  “Hajimete no Ots...   1991-01-01   
145    ts251477      My First Errand   SHOW  “Hajimete no Ots...   1991-01-01   
...         ...                  ...    ...                  ...          ...   
5810  tm1225897           Social Man  MOVIE  Two competitive ...   2021-01-01   
5833   ts307884           HQ Barbers   SHOW  When a family ru...   2021-01-01   
5840  tm1216735      Sun of the Soil  MOVIE  In 14th-century ...   2022-01-01   
5844  tm1074617  Bling Empire - T...  MOVIE  The stars of "Bl...   2021-01-01   
5848  tm1035612  Dad Stop Embarra...  MOVIE  Jamie Foxx, Davi...   2021-01-01   

     age_certification  runtime         genres production_countries  seasons  \
0                TV-MA       51  documentation                   US      1.0   
75                   R       80         comedy                   US      NaN   
145               TV-G       18  documentation                   JP     12.0   
145               TV-G       18         family                   JP     12.0   
145               TV-G       18        reality                   JP     12.0   
...                ...      ...            ...                  ...      ...   
5810               NaN       96          drama                  NaN      NaN   
5833             TV-14       24         comedy                   NG      1.0   
5840               NaN       26            NaN                  NaN      NaN   
5844               NaN       35            NaN                   US      NaN   
5848             PG-13       37            NaN                   US      NaN   

         imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
0            NaN         NaN         NaN            0.600         NaN  
75           NaN         NaN         NaN            3.377         7.5  
145          NaN         NaN         NaN            7.730         7.8  
145          NaN         NaN         NaN            7.730         7.8  
145          NaN         NaN         NaN            7.730         7.8  
...          ...         ...         ...              ...         ...  
5810  tt20198164         NaN         NaN              NaN         NaN  
5833         NaN         NaN         NaN            0.840         NaN  
5840         NaN         NaN         NaN            1.179         7.0  
5844         NaN         NaN         NaN              NaN         NaN  
5848         NaN         NaN         NaN            1.296        10.0  

[842 rows x 15 columns]

In [22]:
#由于没有平分，直接删掉

In [23]:
df1.dropna(subset = ["imdb_score"],inplace = True)

In [24]:
df1["imdb_score"].isnull().sum()

np.int64(0)

In [25]:
df1.query("imdb_score.isnull()")  #可见没有所谓的imdb评分为0的电视剧或电影了

Empty DataFrame
Columns: [id, title, type, description, release_year, age_certification, runtime, genres, production_countries, seasons, imdb_id, imdb_score, imdb_votes, tmdb_popularity, tmdb_score]
Index: []

In [26]:
len(df1.query("genres.isnull()"))   #流派显示空值的有六行

6

In [27]:
df1.query("genres.isnull()")

id                title   type          description release_year  \
1813    ts77824  My Next Guest Ne...   SHOW  TV legend David ...   2018-01-01   
1939   ts215037  Minecraft: Story...   SHOW  MInecraft: Story...   2018-01-01   
2386    ts74805  A Little Help wi...   SHOW  In this unscript...   2018-01-01   
2658   ts265844             #ABtalks   SHOW  #ABtalks is a Yo...   2018-01-01   
4274  tm1172010    The Lockdown Plan  MOVIE                  NaN   2020-01-01   
4648  tm1113921             In Vitro  MOVIE  'In Vitro' is an...   2019-01-01   

     age_certification  runtime genres production_countries  seasons  \
1813             TV-MA       50    NaN                   US      4.0   
1939             TV-PG       52    NaN                   US      1.0   
2386              TV-G       24    NaN                   US      1.0   
2658             TV-PG       68    NaN                  NaN      1.0   
4274               NaN       49    NaN                  NaN      NaN   
4648               NaN       27    NaN                  NaN      NaN   

         imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
1813   tt7829834         7.8      5581.0            8.217         7.6  
1939  tt10498322         5.6       347.0              NaN         NaN  
2386   tt7204366         6.3       237.0            1.621         6.2  
2658  tt12635254         9.6         7.0              NaN         NaN  
4274  tt13079112         6.5         NaN              NaN         NaN  
4648  tt10545994         7.7         NaN              NaN         NaN

In [28]:
df1.dropna(subset=["genres"],inplace = True)

In [29]:
df1["genres"].isnull().sum()

np.int64(0)

In [30]:
df1[ df1["genres"].isnull() ]  #等于  df1.query("genres.isnull()")

Empty DataFrame
Columns: [id, title, type, description, release_year, age_certification, runtime, genres, production_countries, seasons, imdb_id, imdb_score, imdb_votes, tmdb_popularity, tmdb_score]
Index: []

**重复数据**

In [31]:
## 先思考，第一个表df1是表示所有电视剧电影的信息，名字那些重复无所谓呀，因为它有的是根据类别又分了行，而且根据类别分了行
## imdb的分数也不会变呀，分成了两行，imdb分数也会是一样的，只是在这个类别是7分，那个类别也是7分，根据duplicated这个方法
## 对df进行查找重复行是，是找完全相同的行。可以直接使用

In [32]:
df1.duplicated().sum()

np.int64(0)

In [33]:
df2

person_id         id                name            character      role
0          3748    tm84618      Robert De Niro        Travis Bickle     ACTOR
1         14658    tm84618        Jodie Foster        Iris Steensma     ACTOR
2          7064    tm84618       Albert Brooks                  Tom     ACTOR
3          3739    tm84618       Harvey Keitel  Matthew 'Sport' ...     ACTOR
4         48933    tm84618     Cybill Shepherd                Betsy     ACTOR
...         ...        ...                 ...                  ...       ...
77796    736339  tm1059008    Adelaida Buscato            María Paz     ACTOR
77797    399499  tm1059008  Luz Stella Luengas         Karen Bayona     ACTOR
77798    373198  tm1059008         Inés Prieto                Fanny     ACTOR
77799    378132  tm1059008        Isabel Gaona               Cacica     ACTOR
77800   1950416  tm1059008      Julian Gaviria                  NaN  DIRECTOR

[77801 rows x 5 columns]

In [34]:
df2.duplicated().sum()

np.int64(0)

**不一致数据**

In [35]:
df1["genres"].value_counts()  #里面所有流派都没有说重复一个意思的流派

genres
drama            3357
comedy           2419
thriller         1446
action           1339
romance          1080
crime            1066
documentation     981
family            769
animation         732
fantasy           727
european          679
scifi             647
horror            438
history           336
music             266
reality           226
war               221
sport             188
western            53
Name: count, dtype: int64

In [36]:
"" in df1["genres"].values   #再确认里面是不是有漏网之鱼！可恶，谁会用空的字符串来迷惑我！

False

In [37]:
df1["production_countries"].value_counts()  #像这里因为国家太多了，展示不全，我们需要调用`display.max_rows`设置为`None`，就是无上限

production_countries
US    5648
IN    1610
GB    1068
JP    1046
FR     720
      ... 
CU       1
LK       1
GT       1
AF       1
FO       1
Name: count, Length: 108, dtype: int64

但因为我们只是在当前调用`value_counts`时才需要看完整结果，所以可以结合`option_context`，只更改临时上限。

In [38]:
with pd.option_context("display.max_rows",None):  #出来的结果交给deepseek帮忙看吧，忒多了！
    print(df1["production_countries"].value_counts())

production_countries
US         5648
IN         1610
GB         1068
JP         1046
FR          720
ES          637
KR          637
CA          608
DE          383
CN          295
MX          264
IT          224
BR          221
AU          217
TR          195
PH          192
AR          150
ID          149
BE          148
TW          133
NG          131
PL          126
ZA          103
HK          102
NL          102
CO           94
EG           93
DK           89
TH           87
SE           81
LB           70
NO           68
AE           52
IE           49
SG           47
XX           43
IL           42
RU           41
CL           35
CH           33
PS           32
BG           31
MY           30
SA           28
IS           28
AT           28
NZ           27
LU           27
PE           26
RO           25
QA           24
CZ           22
JO           19
FI           18
HU           18
UY           15
MA           15
PT           14
KH           10
KW           10
PR            9
PK 

从以上输出结果来看，出品国家都用两位的国家代码来表示，除了里面存在一个的`Lebanon`值。

`Lebanon`的国家代码是`LB`，出现了39次，说明此处数据不一致。`LB`和`Lebanon`都在表示同一国家，需要进行统一。

In [39]:
df1["production_countries"] = df1["production_countries"].replace({"Lebanon":"LB"})  #.str.replace和series的replace还是有区别的

In [40]:
with pd.option_context("display.max_rows",None):
    print( "" in df1["production_countries"].values)
    print( df1["production_countries"].value_counts())

False
production_countries
US    5648
IN    1610
GB    1068
JP    1046
FR     720
ES     637
KR     637
CA     608
DE     383
CN     295
MX     264
IT     224
BR     221
AU     217
TR     195
PH     192
AR     150
ID     149
BE     148
TW     133
NG     131
PL     126
ZA     103
HK     102
NL     102
CO      94
EG      93
DK      89
TH      87
SE      81
LB      71
NO      68
AE      52
IE      49
SG      47
XX      43
IL      42
RU      41
CL      35
CH      33
PS      32
BG      31
MY      30
SA      28
AT      28
IS      28
NZ      27
LU      27
PE      26
RO      25
QA      24
CZ      22
JO      19
FI      18
HU      18
MA      15
UY      15
PT      14
KW      10
KH      10
PK       9
PR       9
UA       8
VN       8
MT       8
SU       7
CD       7
TN       7
LT       7
IR       7
GH       6
SN       6
AL       6
KE       6
IQ       5
MU       5
CY       5
TZ       4
SY       4
MC       4
IO       4
KN       4
GR       4
BD       3
BS       3
DZ       3
GL       3
AO       3
CM   

In [41]:
# 国家也可能存在不一致，但是这并不影响我们此次的分析目标

In [42]:
df2

person_id         id                name            character      role
0          3748    tm84618      Robert De Niro        Travis Bickle     ACTOR
1         14658    tm84618        Jodie Foster        Iris Steensma     ACTOR
2          7064    tm84618       Albert Brooks                  Tom     ACTOR
3          3739    tm84618       Harvey Keitel  Matthew 'Sport' ...     ACTOR
4         48933    tm84618     Cybill Shepherd                Betsy     ACTOR
...         ...        ...                 ...                  ...       ...
77796    736339  tm1059008    Adelaida Buscato            María Paz     ACTOR
77797    399499  tm1059008  Luz Stella Luengas         Karen Bayona     ACTOR
77798    373198  tm1059008         Inés Prieto                Fanny     ACTOR
77799    378132  tm1059008        Isabel Gaona               Cacica     ACTOR
77800   1950416  tm1059008      Julian Gaviria                  NaN  DIRECTOR

[77801 rows x 5 columns]

In [43]:
df2["role"]   #类型为object，但应转为category，因为变量值有限才两个，字符串占用的空间更大，category可以节约内存空间

0           ACTOR
1           ACTOR
2           ACTOR
3           ACTOR
4           ACTOR
           ...   
77796       ACTOR
77797       ACTOR
77798       ACTOR
77799       ACTOR
77800    DIRECTOR
Name: role, Length: 77801, dtype: object

In [44]:
df2["role"].value_counts()

role
ACTOR       73251
DIRECTOR     4550
Name: count, dtype: int64

In [45]:
# 既然演员信息表中的role为演员和导演两个，那么就是有限数量的变量，我们可以转成category种类

In [46]:
df2["role"] = df2["role"].astype("category")

In [47]:
df2["role"]

0           ACTOR
1           ACTOR
2           ACTOR
3           ACTOR
4           ACTOR
           ...   
77796       ACTOR
77797       ACTOR
77798       ACTOR
77799       ACTOR
77800    DIRECTOR
Name: role, Length: 77801, dtype: category
Categories (2, object): ['ACTOR', 'DIRECTOR']

**处理无效数据**

In [48]:
df1.describe()  #可以看到不存在脱离实际意义的值，都是在正常范围内

release_year       runtime      seasons    imdb_score  \
count                16970  16970.000000  5954.000000  16970.000000   
mean   2015-11-14 22:42...     80.912552     2.455492      6.514207   
min    1954-01-01 00:00:00      0.000000     1.000000      1.500000   
25%    2015-01-01 00:00:00     45.000000     1.000000      5.800000   
50%    2018-01-01 00:00:00     90.000000     2.000000      6.600000   
75%    2020-01-01 00:00:00    107.000000     3.000000      7.300000   
max    2022-01-01 00:00:00    225.000000    42.000000      9.500000   
std                    NaN     39.596172     2.869428      1.131095   

         imdb_votes  tmdb_popularity    tmdb_score  
count  1.694100e+04     16842.000000  16515.000000  
mean   3.281655e+04        29.396307      6.846933  
min    5.000000e+00         0.600000      1.000000  
25%    7.800000e+02         4.070000      6.200000  
50%    3.508000e+03        10.195000      6.900000  
75%    1.697800e+04        23.639000      7.500000  
max    2.294231e+06      2274.044000     10.000000  
std    1.141492e+05        93.178235      1.078831

In [49]:
df2  #df2中的东西全是字符串类型，我们在这个表里重点关注演员名字和影视id，都没有无效的数据

person_id         id                name            character      role
0          3748    tm84618      Robert De Niro        Travis Bickle     ACTOR
1         14658    tm84618        Jodie Foster        Iris Steensma     ACTOR
2          7064    tm84618       Albert Brooks                  Tom     ACTOR
3          3739    tm84618       Harvey Keitel  Matthew 'Sport' ...     ACTOR
4         48933    tm84618     Cybill Shepherd                Betsy     ACTOR
...         ...        ...                 ...                  ...       ...
77796    736339  tm1059008    Adelaida Buscato            María Paz     ACTOR
77797    399499  tm1059008  Luz Stella Luengas         Karen Bayona     ACTOR
77798    373198  tm1059008         Inés Prieto                Fanny     ACTOR
77799    378132  tm1059008        Isabel Gaona               Cacica     ACTOR
77800   1950416  tm1059008      Julian Gaviria                  NaN  DIRECTOR

[77801 rows x 5 columns]

**保存数据**

In [50]:
df1

id                title   type          description release_year  \
1       tm84618          Taxi Driver  MOVIE  A mentally unsta...   1976-01-01   
1       tm84618          Taxi Driver  MOVIE  A mentally unsta...   1976-01-01   
2      tm154986          Deliverance  MOVIE  Intent on seeing...   1972-01-01   
2      tm154986          Deliverance  MOVIE  Intent on seeing...   1972-01-01   
2      tm154986          Deliverance  MOVIE  Intent on seeing...   1972-01-01   
...         ...                  ...    ...                  ...          ...   
5846   tm898842          C/O Kaadhal  MOVIE  A heart warming ...   2021-01-01   
5847  tm1059008              Lokillo  MOVIE  A controversial ...   2021-01-01   
5849   ts271048  Mighty Little Bh...   SHOW  With winter behi...   2021-01-01   
5849   ts271048  Mighty Little Bh...   SHOW  With winter behi...   2021-01-01   
5849   ts271048  Mighty Little Bh...   SHOW  With winter behi...   2021-01-01   

     age_certification  runtime     genres production_countries  seasons  \
1                    R      114      drama                   US      NaN   
1                    R      114      crime                   US      NaN   
2                    R      109      drama                   US      NaN   
2                    R      109     action                   US      NaN   
2                    R      109   thriller                   US      NaN   
...                ...      ...        ...                  ...      ...   
5846               NaN      134      drama                  NaN      NaN   
5847               NaN       90     comedy                   CO      NaN   
5849               NaN        7     family                  NaN      1.0   
5849               NaN        7  animation                  NaN      1.0   
5849               NaN        7     comedy                  NaN      1.0   

         imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
1      tt0075314         8.2    808582.0           40.965       8.179  
1      tt0075314         8.2    808582.0           40.965       8.179  
2      tt0068473         7.7    107673.0           10.010       7.300  
2      tt0068473         7.7    107673.0           10.010       7.300  
2      tt0068473         7.7    107673.0           10.010       7.300  
...          ...         ...         ...              ...         ...  
5846  tt11803618         7.7       348.0              NaN         NaN  
5847  tt14585902         3.8        68.0           26.005       6.300  
5849  tt13711094         7.8        18.0            2.289      10.000  
5849  tt13711094         7.8        18.0            2.289      10.000  
5849  tt13711094         7.8        18.0            2.289      10.000  

[16970 rows x 15 columns]

In [51]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
Index: 16970 entries, 1 to 5849
Data columns (total 15 columns):
 #   Column                Non-Null Count  Dtype         
---  ------                --------------  -----         
 0   id                    16970 non-null  object        
 1   title                 16970 non-null  object        
 2   type                  16970 non-null  object        
 3   description           16965 non-null  object        
 4   release_year          16970 non-null  datetime64[ns]
 5   age_certification     10506 non-null  object        
 6   runtime               16970 non-null  int64         
 7   genres                16970 non-null  object        
 8   production_countries  16670 non-null  object        
 9   seasons               5954 non-null   float64       
 10  imdb_id               16970 non-null  object        
 11  imdb_score            16970 non-null  float64       
 12  imdb_votes            16941 non-null  float64       
 13  tmdb_popularity       

In [52]:
df1.describe()

release_year       runtime      seasons    imdb_score  \
count                16970  16970.000000  5954.000000  16970.000000   
mean   2015-11-14 22:42...     80.912552     2.455492      6.514207   
min    1954-01-01 00:00:00      0.000000     1.000000      1.500000   
25%    2015-01-01 00:00:00     45.000000     1.000000      5.800000   
50%    2018-01-01 00:00:00     90.000000     2.000000      6.600000   
75%    2020-01-01 00:00:00    107.000000     3.000000      7.300000   
max    2022-01-01 00:00:00    225.000000    42.000000      9.500000   
std                    NaN     39.596172     2.869428      1.131095   

         imdb_votes  tmdb_popularity    tmdb_score  
count  1.694100e+04     16842.000000  16515.000000  
mean   3.281655e+04        29.396307      6.846933  
min    5.000000e+00         0.600000      1.000000  
25%    7.800000e+02         4.070000      6.200000  
50%    3.508000e+03        10.195000      6.900000  
75%    1.697800e+04        23.639000      7.500000  
max    2.294231e+06      2274.044000     10.000000  
std    1.141492e+05        93.178235      1.078831

In [53]:
cleaned_title = df1.copy()

In [54]:
df2

person_id         id                name            character      role
0          3748    tm84618      Robert De Niro        Travis Bickle     ACTOR
1         14658    tm84618        Jodie Foster        Iris Steensma     ACTOR
2          7064    tm84618       Albert Brooks                  Tom     ACTOR
3          3739    tm84618       Harvey Keitel  Matthew 'Sport' ...     ACTOR
4         48933    tm84618     Cybill Shepherd                Betsy     ACTOR
...         ...        ...                 ...                  ...       ...
77796    736339  tm1059008    Adelaida Buscato            María Paz     ACTOR
77797    399499  tm1059008  Luz Stella Luengas         Karen Bayona     ACTOR
77798    373198  tm1059008         Inés Prieto                Fanny     ACTOR
77799    378132  tm1059008        Isabel Gaona               Cacica     ACTOR
77800   1950416  tm1059008      Julian Gaviria                  NaN  DIRECTOR

[77801 rows x 5 columns]

In [55]:
df2.describe()

person_id       id                 name character   role
count      77801    77801                77801     68029  77801
unique     54589     5489                54314     47274      2
top        38636  tm32982  Kareena Kapoor Khan      Self  ACTOR
freq          25      208                   25      1950  73251

In [56]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 77801 entries, 0 to 77800
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype   
---  ------     --------------  -----   
 0   person_id  77801 non-null  object  
 1   id         77801 non-null  object  
 2   name       77801 non-null  object  
 3   character  68029 non-null  object  
 4   role       77801 non-null  category
dtypes: category(1), object(4)
memory usage: 2.4+ MB


In [57]:
cleaned_credit = df2.copy()

In [58]:
#  cleaned_title 和 cleaned_credit

# 整理数据

In [59]:
cleaned_title 

id                title   type          description release_year  \
1       tm84618          Taxi Driver  MOVIE  A mentally unsta...   1976-01-01   
1       tm84618          Taxi Driver  MOVIE  A mentally unsta...   1976-01-01   
2      tm154986          Deliverance  MOVIE  Intent on seeing...   1972-01-01   
2      tm154986          Deliverance  MOVIE  Intent on seeing...   1972-01-01   
2      tm154986          Deliverance  MOVIE  Intent on seeing...   1972-01-01   
...         ...                  ...    ...                  ...          ...   
5846   tm898842          C/O Kaadhal  MOVIE  A heart warming ...   2021-01-01   
5847  tm1059008              Lokillo  MOVIE  A controversial ...   2021-01-01   
5849   ts271048  Mighty Little Bh...   SHOW  With winter behi...   2021-01-01   
5849   ts271048  Mighty Little Bh...   SHOW  With winter behi...   2021-01-01   
5849   ts271048  Mighty Little Bh...   SHOW  With winter behi...   2021-01-01   

     age_certification  runtime     genres production_countries  seasons  \
1                    R      114      drama                   US      NaN   
1                    R      114      crime                   US      NaN   
2                    R      109      drama                   US      NaN   
2                    R      109     action                   US      NaN   
2                    R      109   thriller                   US      NaN   
...                ...      ...        ...                  ...      ...   
5846               NaN      134      drama                  NaN      NaN   
5847               NaN       90     comedy                   CO      NaN   
5849               NaN        7     family                  NaN      1.0   
5849               NaN        7  animation                  NaN      1.0   
5849               NaN        7     comedy                  NaN      1.0   

         imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
1      tt0075314         8.2    808582.0           40.965       8.179  
1      tt0075314         8.2    808582.0           40.965       8.179  
2      tt0068473         7.7    107673.0           10.010       7.300  
2      tt0068473         7.7    107673.0           10.010       7.300  
2      tt0068473         7.7    107673.0           10.010       7.300  
...          ...         ...         ...              ...         ...  
5846  tt11803618         7.7       348.0              NaN         NaN  
5847  tt14585902         3.8        68.0           26.005       6.300  
5849  tt13711094         7.8        18.0            2.289      10.000  
5849  tt13711094         7.8        18.0            2.289      10.000  
5849  tt13711094         7.8        18.0            2.289      10.000  

[16970 rows x 15 columns]

In [60]:
cleaned_credit

person_id         id                name            character      role
0          3748    tm84618      Robert De Niro        Travis Bickle     ACTOR
1         14658    tm84618        Jodie Foster        Iris Steensma     ACTOR
2          7064    tm84618       Albert Brooks                  Tom     ACTOR
3          3739    tm84618       Harvey Keitel  Matthew 'Sport' ...     ACTOR
4         48933    tm84618     Cybill Shepherd                Betsy     ACTOR
...         ...        ...                 ...                  ...       ...
77796    736339  tm1059008    Adelaida Buscato            María Paz     ACTOR
77797    399499  tm1059008  Luz Stella Luengas         Karen Bayona     ACTOR
77798    373198  tm1059008         Inés Prieto                Fanny     ACTOR
77799    378132  tm1059008        Isabel Gaona               Cacica     ACTOR
77800   1950416  tm1059008      Julian Gaviria                  NaN  DIRECTOR

[77801 rows x 5 columns]

**这里我们应该思考，因为我们目标是找流派中高分的演员，我们目前有两个表格|**

**应该要通过影视作品的id去进行匹配合并，必须是两者都有的id才行，不然比如电影表里有但是演员表没有的影视作品缺少了演员**

**又或者演员表里有的影视作品表里没有这个作品那缺少流派，也是影响我们分析的，我们合并采用的方法应该是默认的how= “inner”**

In [61]:
credit_title = pd.merge(cleaned_credit,cleaned_title,on = "id",how ="inner" )

In [62]:
credit_title

person_id         id                name      character      role  \
0           3748    tm84618      Robert De Niro  Travis Bickle     ACTOR   
1           3748    tm84618      Robert De Niro  Travis Bickle     ACTOR   
2          14658    tm84618        Jodie Foster  Iris Steensma     ACTOR   
3          14658    tm84618        Jodie Foster  Iris Steensma     ACTOR   
4           7064    tm84618       Albert Brooks            Tom     ACTOR   
...          ...        ...                 ...            ...       ...   
276104    736339  tm1059008    Adelaida Buscato      María Paz     ACTOR   
276105    399499  tm1059008  Luz Stella Luengas   Karen Bayona     ACTOR   
276106    373198  tm1059008         Inés Prieto          Fanny     ACTOR   
276107    378132  tm1059008        Isabel Gaona         Cacica     ACTOR   
276108   1950416  tm1059008      Julian Gaviria            NaN  DIRECTOR   

              title   type          description release_year  \
0       Taxi Driver  MOVIE  A mentally unsta...   1976-01-01   
1       Taxi Driver  MOVIE  A mentally unsta...   1976-01-01   
2       Taxi Driver  MOVIE  A mentally unsta...   1976-01-01   
3       Taxi Driver  MOVIE  A mentally unsta...   1976-01-01   
4       Taxi Driver  MOVIE  A mentally unsta...   1976-01-01   
...             ...    ...                  ...          ...   
276104      Lokillo  MOVIE  A controversial ...   2021-01-01   
276105      Lokillo  MOVIE  A controversial ...   2021-01-01   
276106      Lokillo  MOVIE  A controversial ...   2021-01-01   
276107      Lokillo  MOVIE  A controversial ...   2021-01-01   
276108      Lokillo  MOVIE  A controversial ...   2021-01-01   

       age_certification  runtime  genres production_countries  seasons  \
0                      R      114   drama                   US      NaN   
1                      R      114   crime                   US      NaN   
2                      R      114   drama                   US      NaN   
3                      R      114   crime                   US      NaN   
4                      R      114   drama                   US      NaN   
...                  ...      ...     ...                  ...      ...   
276104               NaN       90  comedy                   CO      NaN   
276105               NaN       90  comedy                   CO      NaN   
276106               NaN       90  comedy                   CO      NaN   
276107               NaN       90  comedy                   CO      NaN   
276108               NaN       90  comedy                   CO      NaN   

           imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
0        tt0075314         8.2    808582.0           40.965       8.179  
1        tt0075314         8.2    808582.0           40.965       8.179  
2        tt0075314         8.2    808582.0           40.965       8.179  
3        tt0075314         8.2    808582.0           40.965       8.179  
4        tt0075314         8.2    808582.0           40.965       8.179  
...            ...         ...         ...              ...         ...  
276104  tt14585902         3.8        68.0           26.005       6.300  
276105  tt14585902         3.8        68.0           26.005       6.300  
276106  tt14585902         3.8        68.0           26.005       6.300  
276107  tt14585902         3.8        68.0           26.005       6.300  
276108  tt14585902         3.8        68.0           26.005       6.300  

[276109 rows x 19 columns]

In [63]:
credit_title = credit_title.query("role =='ACTOR'") #其实这一步应该在清理df2无效数据那应该得做了

In [64]:
# 一个演员（一个perosnid）可能出演很多部喜剧片，或动作片或其他，出演了很多部某个流派中的许多电影，都有评分，肯定是都求平均分呀，然后平均分比高低

In [65]:
credit_title.query("person_id == '1000'")  #可以看到1000编号这一个演员出演了很多部drama类的，肯定是求drama类这些影片他获得的平均分呀

person_id        id          name            character   role  \
12341       1000  tm192199  Martin Sheen           Jason Wynn  ACTOR   
12342       1000  tm192199  Martin Sheen           Jason Wynn  ACTOR   
12343       1000  tm192199  Martin Sheen           Jason Wynn  ACTOR   
12344       1000  tm192199  Martin Sheen           Jason Wynn  ACTOR   
16240       1000  tm111828  Martin Sheen         Roger Strong  ACTOR   
16241       1000  tm111828  Martin Sheen         Roger Strong  ACTOR   
17239       1000   tm27911  Martin Sheen  Capt. Oliver Cha...  ACTOR   
17240       1000   tm27911  Martin Sheen  Capt. Oliver Cha...  ACTOR   
17241       1000   tm27911  Martin Sheen  Capt. Oliver Cha...  ACTOR   
17242       1000   tm27911  Martin Sheen  Capt. Oliver Cha...  ACTOR   
47598       1000   tm44730  Martin Sheen            Uncle Ben  ACTOR   
47599       1000   tm44730  Martin Sheen            Uncle Ben  ACTOR   
47600       1000   tm44730  Martin Sheen            Uncle Ben  ACTOR   
80822       1000   ts35251  Martin Sheen        Robert Hanson  ACTOR   
80823       1000   ts35251  Martin Sheen        Robert Hanson  ACTOR   
145934      1000  tm366820  Martin Sheen         Oral Roberts  ACTOR   

                      title   type          description release_year  \
12341                 Spawn  MOVIE  After being murd...   1997-01-01   
12342                 Spawn  MOVIE  After being murd...   1997-01-01   
12343                 Spawn  MOVIE  After being murd...   1997-01-01   
12344                 Spawn  MOVIE  After being murd...   1997-01-01   
16240   Catch Me If You Can  MOVIE  A true story abo...   2002-01-01   
16241   Catch Me If You Can  MOVIE  A true story abo...   2002-01-01   
17239          The Departed  MOVIE  To take down Sou...   2006-01-01   
17240          The Departed  MOVIE  To take down Sou...   2006-01-01   
17241          The Departed  MOVIE  To take down Sou...   2006-01-01   
17242          The Departed  MOVIE  To take down Sou...   2006-01-01   
47598   The Amazing Spid...  MOVIE  Peter Parker is ...   2012-01-01   
47599   The Amazing Spid...  MOVIE  Peter Parker is ...   2012-01-01   
47600   The Amazing Spid...  MOVIE  Peter Parker is ...   2012-01-01   
80822     Grace and Frankie   SHOW  Elegant, proper ...   2015-01-01   
80823     Grace and Frankie   SHOW  Elegant, proper ...   2015-01-01   
145934          Come Sunday  MOVIE  Evangelist Carlt...   2018-01-01   

       age_certification  runtime    genres production_countries  seasons  \
12341              PG-13       96     scifi                   US      NaN   
12342              PG-13       96    action                   US      NaN   
12343              PG-13       96     drama                   US      NaN   
12344              PG-13       96    horror                   US      NaN   
16240              PG-13      141     drama                   US      NaN   
16241              PG-13      141     crime                   US      NaN   
17239                  R      151     drama                   US      NaN   
17240                  R      151  thriller                   US      NaN   
17241                  R      151     crime                   US      NaN   
17242                  R      151    action                   US      NaN   
47598              PG-13      136    action                   US      NaN   
47599              PG-13      136   fantasy                   US      NaN   
47600              PG-13      136     scifi                   US      NaN   
80822              TV-MA       29     drama                   US      7.0   
80823              TV-MA       29    comedy                   US      7.0   
145934               NaN      106     drama                   US      NaN   

          imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
12341   tt0120177         5.2     68184.0           16.989       5.343  
12342   tt0120177         5.2     68184.0           16.989       5.343  
12343   tt0120177         5.

In [66]:
pd.pivot_table(credit_title,index =["genres","person_id"],values = "imdb_score",aggfunc = np.mean)   #这里因为表格里没有合适的columns我就不指定了

C:\Users\Windows 10\AppData\Local\Temp\ipykernel_10808\1213162902.py:1: FutureWarning: The provided callable <function mean at 0x000001C0FDE81C60> is currently using DataFrameGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "mean" instead.
  pd.pivot_table(credit_title,index =["genres","person_id"],values = "imdb_score",aggfunc = np.mean)   #这里因为表格里没有合适的columns我就不指定了


imdb_score
genres  person_id            
action  1000         6.866667
        100007       7.000000
        100013       6.400000
        100019       6.500000
        100020       6.500000
...                       ...
western 993735       6.500000
        998673       7.300000
        998674       7.300000
        998675       7.300000
        99940        4.000000

[168881 rows x 1 columns]

In [67]:
# 所以分组逻辑就是，有哪些流派是吧动作片喜剧片惊悚片等等，然后这个流派下又有哪些演员演过？然后里面计算imdb的平均分是多少

In [68]:
imdbscore_groupby = pd.pivot_table(credit_title,index =["genres","person_id"],values = "imdb_score",aggfunc = np.mean)  

C:\Users\Windows 10\AppData\Local\Temp\ipykernel_10808\4060119275.py:1: FutureWarning: The provided callable <function mean at 0x000001C0FDE81C60> is currently using DataFrameGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "mean" instead.
  imdbscore_groupby = pd.pivot_table(credit_title,index =["genres","person_id"],values = "imdb_score",aggfunc = np.mean)


In [69]:
imdbscore_groupby

imdb_score
genres  person_id            
action  1000         6.866667
        100007       7.000000
        100013       6.400000
        100019       6.500000
        100020       6.500000
...                       ...
western 993735       6.500000
        998673       7.300000
        998674       7.300000
        998675       7.300000
        99940        4.000000

[168881 rows x 1 columns]

In [70]:
reset_imdbscore_groupby = imdbscore_groupby.reset_index()

In [71]:
reset_imdbscore_groupby 

genres person_id  imdb_score
0        action      1000    6.866667
1        action    100007    7.000000
2        action    100013    6.400000
3        action    100019    6.500000
4        action    100020    6.500000
...         ...       ...         ...
168876  western    993735    6.500000
168877  western    998673    7.300000
168878  western    998674    7.300000
168879  western    998675    7.300000
168880  western     99940    4.000000

[168881 rows x 3 columns]

In [72]:
genres_max_scores = reset_imdbscore_groupby.groupby("genres")["imdb_score"].max()

In [73]:
genres_max_scores

genres
action           9.3
animation        9.3
comedy           9.2
crime            9.5
documentation    9.1
drama            9.5
european         8.9
family           9.3
fantasy          9.3
history          9.1
horror           9.0
music            8.8
reality          8.9
romance          9.2
scifi            9.3
sport            9.1
thriller         9.5
war              8.8
western          8.9
Name: imdb_score, dtype: float64

In [74]:
genres_max_score_with_person_id = pd.merge(reset_imdbscore_groupby,genres_max_scores,on = ["genres","imdb_score"])

In [75]:
genres_max_score_with_person_id

genres person_id  imdb_score
0     action     12790         9.3
1     action      1303         9.3
2     action     21033         9.3
3     action    336830         9.3
4     action     86591         9.3
..       ...       ...         ...
131      war    826547         8.8
132  western     22311         8.9
133  western     28166         8.9
134  western     28180         8.9
135  western     93017         8.9

[136 rows x 3 columns]

In [76]:
actor_id_with_names = cleaned_credit[["person_id","name"]].drop_duplicates()  #提取出来的df再使用删除重复行

In [77]:
actor_id_with_names.head(10)

person_id             name
0      3748   Robert De Niro
1     14658     Jodie Foster
2      7064    Albert Brooks
3      3739    Harvey Keitel
4     48933  Cybill Shepherd
5     32267      Peter Boyle
6    519612   Leonard Harris
7     29068   Diahnne Abbott
8    519613      Gino Ardito
9      3308  Martin Scorsese

In [78]:
genres_max_score_with_actor_name = pd.merge(genres_max_score_with_person_id,actor_id_with_names,on = "person_id")

In [79]:
genres_max_score_with_actor_name

genres person_id  imdb_score                name
0     action     12790         9.3         Olivia Hack
1     action      1303         9.3       Jessie Flower
2     action     21033         9.3          Zach Tyler
3     action    336830         9.3     André Sogliuzzo
4     action     86591         9.3       Cricket Leigh
..       ...       ...         ...                 ...
131      war    826547         8.8         Yuto Uemura
132  western     22311         8.9     Koichi Yamadera
133  western     28166         8.9  Megumi Hayashibara
134  western     28180         8.9      Unsho Ishizuka
135  western     93017         8.9            Aoi Tada

[136 rows x 4 columns]

为了把相同流派都排序在一起，我们还可以用`sort_values`方法，把结果里面的行根据`genres`进行排序，然后用`reset_index`把索引重新排序。

索引重新排序后，DataFrame会多出`index`一列，我们可以再把`index`列进行删除。

In [80]:
genres_max_score_with_actor_name = genres_max_score_with_actor_name.sort_values("genres").reset_index().drop("index", axis=1)
genres_max_score_with_actor_name

genres person_id  imdb_score                name
0     action     12790         9.3         Olivia Hack
1     action      1303         9.3       Jessie Flower
2     action     21033         9.3          Zach Tyler
3     action    336830         9.3     André Sogliuzzo
4     action     86591         9.3       Cricket Leigh
..       ...       ...         ...                 ...
131      war    826547         8.8         Yuto Uemura
132  western     28180         8.9      Unsho Ishizuka
133  western     22311         8.9     Koichi Yamadera
134  western     28166         8.9  Megumi Hayashibara
135  western     93017         8.9            Aoi Tada

[136 rows x 4 columns]